### DISCUSSION TITLE
Unbundle pages in a separate app (Monorepo)

### DISCUSSION THREAD

**Discussion by NiklasBuchfink** [vikejs/vike#1202] [Discussion] [answered]
## What we want to achieve
We have at inlang an NX monorepo with a vike website that includes one large client app. Now we want to unbundle the client app in its own folder inside the monorepo to avoid shared components and layouts with a lot of conditional logic, but still want to use vike routing for the client app.

How can we achieve this or what else would you recommend us to do?

### Additional information

We tried to make this structure work, but the routing of the editor and website conflicts.

```
.
└── source-code
    ├── editor (client app)
        ├── src
            ├── pages
            ├── renderer
    ├── website (SSR)
        ├── src
            ├── pages
            ├── renderer
    └── server (Express server)
```

Routes:
- `/editor/*`: client app is served via the editor folder
- `*`: other routes are served via the website folder

**Comment by brillout** [comment]
> we want to unbundle

Why?

> the routing of the editor and website conflicts

What do you mean exactly with "conflicts"? Do you get an error?

If you follow https://vike.dev/file-structure#domain-driven then that file structure should work.

**Reply by NiklasBuchfink** [reply]
> Why?

The main reason is to have no dependencies between our editor and the rest of the website. Changes in the editor should have no impact on the website. This way the editor team and the website team shouldn't conflict with each other in the future.

Do you disagree that it makes sense to unbundle the editor from a website in this scenario?

> What do you mean exactly with "conflicts"? Do you get an error?

If I use two `renderPage` functions with the structure described above to manage the URL requests from our Express router:
1. Only the page directory of the first imported configuration will be used for routing. The other is ignored.
2. It always uses the same renderers

This results in a check whether the route exists in the website folder and the editor folder is rendered. As no page exists twice, nothing is rendered correctly.

**Reply by brillout** [reply]
I see, makes sense.

I recommend to follow the domain-driven file structure setup. (You could as well setup two completely different apps but I don't think the benefits are worth the added complexity in terms of setup.)

**Reply by brillout** [reply]
As for your issue, something seems to be broken about your setup. Make sure to follow the domain-driven file structure example. Or provide a minimal reproduction and I'll have a look at it.

**Reply by NiklasBuchfink** [reply]
Here is a minimal reproduction of what we are trying to achieve:
https://stackblitz.com/edit/stackblitz-starters-bjn9qd

Only the page inside the website folder is working. `/editor` is requested from the editor router, but it only uses the domain-driven file structure of the website folder and therefore doesn't return the editor page.

**Reply by brillout** [reply]
I have difficulties understanding both your problem and what you want. You have two different `vite.config.js` which obviously won't work - which one do you expect Vite to pick?

**Reply by samuelstroschein** [reply]
@brillout two separate apps that should run under the same domain for easier deployment (only one server, DNS, etc. required)

```
google.com <- one app
google.com/maps <- another app

inlang.com <- website app
inlang.com/editor <- editor app
```

Given that Vike is an (express) middleware, using two middleware is expected to work fine. However, the routing does not work as expected. The `/editor` route doesn't resolve. 

1. Vite builds just fine (don't know what the issue with multiple vite configs is supposed to be)
2. Vike seems to have issues resolving the route 

https://github.com/vikejs/vike/assets/35429197/0ca16236-9e4a-4fdb-b219-01c35d652afd

PS the reproduction is missing a key detail: multiple package.json's. We want complete separation, including dependencies. Using domain-driven development with one huge package.json is out of question.

**Reply by brillout** [reply]
If you use two completely different apps/setups then they won't interact with each other, so I don't expect any issues here (except maybe for `$ vite build` but that doesn't seem to be the issue you're having).

I'm guessing you're doing something weird. (Like in your "reproduction" where you have two vite.config.js for a single package.json which obviously won't work.)

I doubt it's related to Vike; it's most likely a user land issue.

**Reply by samuelstroschein** [reply]
Hmm the routing seems odd, it might be an express/usage issue, see the [3 min video](https://www.loom.com/share/7ea072d4f6cc408191be1ff946d6e9e5?sid=49419038-bd4d-418b-901b-bb205722d182) and updated package reproduction which uses individual package.json's, rules out vite as cause etc https://stackblitz.com/edit/stackblitz-starters-fd54nc. 

I think we gonna go for subdomain deployment to make our life easier.

**Reply by samuelstroschein** [reply]
EDIT: given that the express router matches but vike always resolves with a 404, this seems to be a bug in vike

**Reply by brillout** [reply]
If you have a ~~fairly~~ high confidence it's a Vike bug, then provide a proper minimal reproduction:
 - Minimal: only one vite.config.js, one express.js server, etc.
 - No weird stuff that won't work for obvious reasons ;)
 - No video reproduction
 - Publish on GitHub

Thanks! This makes my life a *lot* easier. (Managing my time is increasingly the bottleneck for the whole Vike project.)

**Reply by samuelstroschein** [reply, marked as answer]
https://github.com/vikejs/vike/issues/1212

> - Minimal: only one vite.config.js, one express.js server, etc.

A monorepo can have N vite apps (configs) and N express servers. Our problem is precisely that we want N vite apps (configs), a basic monorepo requirement.

**Reply by brillout** [reply]
Thanks for the reproduction.

A proper reproduction is worth more than 1000 words, images, and videos 🙂
